### PR TITLE
Add `TSLexer::force_serialize_state`

### PR DESCRIPTION
**Pull request by ben-clayton**
A new lexer flag to ensure that an external scanner's state is serialized even when the `scan` function returns false.

This is useful for scanners that perform lookahead scanning, but do not want to override the internal scanner for the current token.

Issue: #2063

### CODE REVIEW

File: lib/src/parser.c [hunk 1]
```diff
@@ -424,14 +424,15 @@ static Subtree ts_parser__lex(
...
+      self->lexer.data.force_serialize_state = false;
       bool found_token = self->language->external_scanner.scan(
         self->external_scanner_payload,
         &self->lexer.data,
         valid_external_tokens
       );
       ts_lexer_finish(&self->lexer, &lookahead_end_byte);
 
-      if (found_token) {
+      if (found_token || self->lexer.data.force_serialize_state) {
```
Comment (ben-clayton) [inline lib/src/parser.c:435]: Note: `ts_parser__lex` is a complex beast, and I'm far from certain that this is perfect. Tests seem happy with this - but I'm more than willing to take any suggestions here.

**Comment by maxbrunsfeld** [comment]
I don’t think incremental parsing could work correctly with this behavior. It’s an important invariant that the scanner’s state can be assumed to be fixed unless external tokens are produced and used in the tree.

**Review by maxbrunsfeld — CHANGES_REQUESTED** [review]
I definitely appreciate the good work that you put into this, especially the integration test that you added. But I don't think we want this option, because I'm pretty sure that it's incompatible with the way that Tree-sitter performs incremental parsing in the *presence* of external scanners.

When reusing a subtree during incremental parsing, we need to make sure the external scanner is in the same state as it was before that subtree was originally produced, and then we need to *reset* the scanner's state to match the state *after* that subtree. Both of those steps rely on the external scanner's states being reflected in the syntax tree itself. That is one important purpose of the serialization.

In the broad sense, I'm open to improvements to this model, but we would need to reason very carefully about how they work with incremental parsing.

**Comment by ben-clayton** [comment]
Thanks, and no worries. I suspected this would be more subtle and difficult than I understood.

**Pull request closed without merging**